### ISSUE TITLE
Support overriding default targetRevision at the AppProject level

### ISSUE THREAD

**Issue by jshearer** [argoproj/argo-cd#2554]
# Motivation

My goal is to support deploying to dev/test and prod clusters from a single repo using merge requests. 

Each `cluster/*` branch will contain all of the manifests to be deployed to that cluster in an app-of-apps style setup. Whenever a build passes, it will commit the correct changes to the `cluster/dev` branch in the manifest repo, and from there the Argo instance running in the dev cluster will sync the changes. 

Eventually, these changes will be promoted to the next environment (maybe `cluster/test`, maybe just :ship: it straight to `cluster/production`) via merge request from `cluster/dev` into `cluster/test` for example. From there, the Argo instance running in the test cluster will pick up the changes, and away we go.

The fundamental problem that I kept coming up against is that in order for each Argo instance to pick up the `Application`s for its corresponding cluster, those Applications must have `targetRevision` set to the correct branch: `cluster/dev` for the dev cluster, `cluster/test` for the test cluster, etc. 

This makes the simple workflow of merging branches to promote Applications a non-starter, since as part of the merge, you would have to manually go in and change each `targetRevision`. One solution, having the CI do it, seems like the CIOps antipattern, which I would love to avoid if possible.

# Proposal

As it stands right now, the _default_ behavior for `targetRevision` if not explicitly specified, is to pick the commit at the HEAD of the repo.

My proposal would be to somehow allow overriding this _default_ behavior. So for example, one place I had imagined this could go is in the `sourceRepos` section of an AppProject. If I could change the default behavior for `targetRevision` to be, an arbitrary ref instead of HEAD for all apps underneath the `services` AppProject, then I could simply configure the `services` AppProject on each cluster to default to the correct `targetRevision`, and then by leaving `targetRevision` out of my `Application` specs, I could get the intended behaviour of merge request based multi-cluster deployments.

---

I have done a bunch of searching for some sort of feature that would give me what I want here, and haven't been successful, hence this issue. That being said, the last bit of this comment by @jon-walton helped me solidify my idea: https://github.com/argoproj/argo-cd/issues/1481#issuecomment-485786784

**Comment by alexec** [comment]
Have you thought about using the branch name for the target revision ?

**Comment by jshearer** [comment]
Yes, that's what I'm proposing here: allow setting a Project-level branch to use, effectively. The issue is the following: if I have a bunch of Applications, each of them has their own individual target revision. When I deploy these apps using an app of apps (which itself has its own target revision), the Application manifests are pulled/generated from my git repo, meaning that the target revision for each Application as committed in the repo needs to match the branch it's being deployed from. So when I want to go and, for example, merge some changes from the branch `cluster/dev` to `cluster/prod`, part of what will need to be changed is the target revision for _each Application_, so that when it gets deployed on the prod cluster, the target revision on the branch `cluster/prod` is `cluster/prod`, not `cluster/dev`.

If I could instead just _not specify_ the target revision on each individual Application, and instead use the one specified at the Project level, this would allow deploying different branches on different clusters without manually changing the target revision multiple times each time you merge between branches.

Does that make any sense? In writing it I realize it's sort of hard to articulate what I mean.

**Comment by jshearer** [comment]
@alexec I could write up a PR for it but I'm not sure if this is the path Argo wants to take -- it's just one possible solution to the branch-per-cluster workflow. Thoughts?

**Comment by jshearer** [comment]
For example, another possible solution that actually feels less smelly is to give a Repository a default ref to pull from. Once https://github.com/argoproj/argo-cd/issues/2251 happens, and repositories exist at the Project layer, then this would also solve the problem of each Application having to specify the target branch for its chosen repo explicitly.

**Comment by alexec** [comment]
Hi @jshearer  - thank you for answering - and the short answer is that we would probably not want to do this. I'm in the process of writing up some recommendations around tracking strategies. The short answer is likely to be you should use tags rather than branches, and tag version you want on different environments when you're ready to promote an app.

**Comment by jshearer** [comment]
Hey @alexec thanks for getting back to me :)

Tags suffer from the same problem as branches, namely that each Application will still need to have it's target tag embedded into it's spec. This means that when you want to deploy an application to the prod cluster, not only would you need to change the `cluster/prod` tag (for example), but you'd also need to _rewrite_ the spec for each Application so that it's target revision points to `cluster/prod`.

**Comment by knkarthik** [comment]
@jshearer Wondering if you've found a solution to this. We want to give our devs the ability to deploy to different dev braches but hard-coding of targetRevision makes this impossible.

**Comment by jshearer** [comment]
... wow, I remember writing this question years ago. Haha not only have we
moved on from argo, we moved on from k8s entirely: it ended up not being
worth the complexity, though it sure was "fun". Good luck to you though! :)

**Comment by prein** [comment]
I'm not sure if we share the same goals, but I'm looking for a solution to testing PRs without allowing overrides in the UI.
Ideally some kind of integration allowing for creating ephemeral preview app deployments or at least updating the targetRevision in a given environment when a PR is open - can't think of how to support multiple PR with different resolutions though. Came across [this](https://codefresh.io/continuous-deployment/creating-temporary-preview-environments-based-pull-requests-argo-cd-codefresh/) but it seems a slight overkill to me.  
What might be good enough - more granular RBAC so I can allow overrides on only targetRevision and nothing else.

**Comment by gajus** [comment]
This is badly needed and I cannot really think of a workaround.

**Comment by blakepettersson** [comment]
If I understand correctly you want to be able to override the default `targetRevision` based upon environment, and/or have the ability to override on a cluster-by-cluster basis. This is possible with [ApplicationSets](https://argo-cd.readthedocs.io/en/stable/user-guide/application-set/).

For example this ~monstrosity~ snippet would enable an application to set its `targetRevision` based upon the cluster annotation
`foo.com/custom-external-dns-version`. If that's not present then we set the `targetRevision` defaults based upon the cluster environment, which is itself defined by a cluster annotation `foo.com/environment`, which in the example below can be one of `dev`, `staging`,  `prod`. 

```
targetRevision: '{{ default (index (dict "dev" "1.0.4" "staging" "1.0.4" "prod" "1.0.4")  (index .metadata.annotations "foo.com/environment")) (index .metadata.annotations "foo.com/custom-external-dns-version") }}'
```

**Comment by good92** [comment]
terraform-aws-eks-blueprints Open: https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1587

**Comment by Lujeni** [comment]
> Wondering if you've found a solution to this. We want to give our devs the ability to deploy to different dev braches but hard-coding of targetRevision makes this impossible.**

You can take a look with `ApplicationSet` + `Git Operator`.
You create a temporary application for every directory, and the template use the `targetRevision: '{{path.basename}}'`
This action is trigger when the dev ask "deploy branch X on developement environement".
Drawbacks, you must automate create/cleanup of this temporary directory and your main version (master) is desync  

### Architecture with a mono repo
```
.../<project-name/versions/master
.../<project-name/versions/issue_1
.../<project-name/versions/issue_2
 ```
 
 ```yaml
---
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: argocd-tester
  namespace: argocd
spec:
  generators:
  - git:
      repoURL: argocd-manager.git
      revision: master
      directories:
      - path: applications/argocd-tester/versions/*
  template:
    metadata:
      name: argocd-tester-'{{path[0]}}'
      namespace: argocd
    spec:
      destination:
        server: 'https://kubernetes.default.svc'
      sources:
        - repoURL: argocd-manager.git
          targetRevision: '{{path.basename}}'
          ref: argocd-manager
          helm:
            valueFiles:
              - {{path}}/default.yaml
              - {{path}}/development.yaml
```

**Comment by andrii-korotkov-verkada** [comment]
Closed since there's been quite a long time since updates and it's unclear if this is still relevant. Feel free to re-open/re-submit the issue if still relevant.
